snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java

```diff
@@ -0,0 +1,237 @@
+    this.snowflakeClient = snowflakeClient;
+    this.fileIO = fileIO;
+    this.catalogProperties = properties;
+    this.closeableGroup = new CloseableGroup();
+    closeableGroup.addCloseable(snowflakeClient);
+    closeableGroup.addCloseable(fileIO);
+    closeableGroup.setSuppressCloseFailure(true);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (null != closeableGroup) {
+      closeableGroup.close();
+    }
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    throw new UnsupportedOperationException(
+        String.format("createNamespace not supported; attempted for namespace '%s'", namespace));
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    LOG.debug("listNamespaces with namespace: {}", namespace);
+    SnowflakeIdentifier scope = NamespaceHelpers.getSnowflakeIdentifierForNamespace(namespace);
+    Preconditions.checkArgument(
+        scope.getType() == SnowflakeIdentifier.Type.ROOT
+            || scope.getType() == SnowflakeIdentifier.Type.DATABASE,
+        "listNamespaces must be at either ROOT or DATABASE level; got %s from namespace %s",
+        scope,
+        namespace);
+    List<SnowflakeIdentifier> sfSchemas = snowflakeClient.listSchemas(scope);
+
+    List<Namespace> namespaceList =
+        sfSchemas.stream()
+            .map(
+                schema -> {
+                  Preconditions.checkState(
+                      schema.getType() == SnowflakeIdentifier.Type.SCHEMA,
+                      "Got identifier of type %s from listSchemas for %s",
+                      schema.getType(),
+                      namespace);
+                  return Namespace.of(schema.getDatabaseName(), schema.getSchemaName());
+                })
+            .collect(Collectors.toList());
+    return namespaceList;
+  }
+
+  @Override
+  public Map<String, String> loadNamespaceMetadata(Namespace namespace)
+      throws NoSuchNamespaceException {
+    LOG.debug("loadNamespaceMetadata with namespace: {}", namespace);
+    return ImmutableMap.of();
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    throw new UnsupportedOperationException(
+        String.format("dropNamespace not supported; attempted for namespace '%s'", namespace));
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) {
+    throw new UnsupportedOperationException(
+        String.format("setProperties not supported; attempted for namespace '%s'", namespace));
+  }
+
+  @Override
+  public boolean removeProperties(Namespace namespace, Set<String> properties) {
+    throw new UnsupportedOperationException(
+        String.format("removeProperties not supported; attempted for namespace '%s'", namespace));
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new SnowflakeTableOperations(
```

Comment:
Each table should be loaded with its own FileIO instance

Reply:
I'm not sure that I agree here. You can't assume that tables share a catalog-level FileIO, but the tables don't necessarily need a separate instance. We reuse the FileIO in other catalogs if the config doesn't change based on the table.

Reply:
Reusing the FileIO will cause problems in certain cases though, so I feel like we shouldn't introduce those here even if there are other catalogs that currently use that approach.  For example, if you want to use S3FileIO (or any other native FileIO), then you'll be limited to a single bucket.  Once the FileIO is initialized, it will only be able to load the table metadata if it was the same as what the catalog was initialized with.  

In Snowflake you can register multiple buckets and arbitrarily point tables to different locations, so we don't want to limit what can be addressed by reusing the same FileIO.  HadoopFileIO is relying on Hadoop's implementation of abstracting the access across buckets to enable this to work, but we shouldn't rely on Hadoop's behavior here.

Reply:
Interesting, I wonder if this would also be a bigger problem down the road due to `io()` being a TableOperations-level method, in case tables themselves start having files in multiple buckets. It seems ideally either FIleIOs are always a URI-authority-level "resolving" FileIO the same way Hadoop's `FileSystem.get` or `Path.getFileSystem` are based on URI authority, or we'd want TableOperations to normally deal with one level higher abstraction where they'd supply some kind of FileIOFactory that can be used to resolve different FileIO instances even if, e.g. metadata.json/manifest-lists/manifests are in different buckets or even different cloud providers.

For now I just pulled out a SnowflakeCatalog.FileIOFactory placeholder to maintain test-injectability and made it instantiate a new FileIO for every call to newTableOps.

Reply:
Eventually we would probably update the underlying FileIO to address the cross bucket issues, but that isn't particularly common for a single table at this point.